matching-core/src/main/java/com/graphhopper/matching/MapMatching.java

```diff
@@ -161,84 +166,201 @@ public MatchResult doWork(List<GPXEntry> gpxList) {
                     + gpxList.size() + "). Correct format?");
         }
 
+        // filter the entries:
+        List<GPXEntry> filteredGPXEntries = filterGPXEntries(gpxList);
+        if (filteredGPXEntries.size() < 2) {
+            throw new IllegalStateException("Only " + filteredGPXEntries.size()
+                    + " filtered GPX entries (from " + gpxList.size()
+                    + "), but two or more are needed");
+        }
+
+        // now find each of the entries in the graph:
         final EdgeFilter edgeFilter = new DefaultEdgeFilter(algoOptions.getWeighting().getFlagEncoder());
 
-        // Compute all candidates first.
-        // TODO: Generate candidates on-the-fly within computeViterbiSequence() if this does not
-        // degrade performance.
-        final List<QueryResult> allCandidates = new ArrayList<>();
-        List<TimeStep<GPXExtension, GPXEntry, Path>> timeSteps = createTimeSteps(gpxList,
-                edgeFilter, allCandidates);
+        List<List<QueryResult>> queriesPerEntry = findGPXEntriesInGraph(filteredGPXEntries, edgeFilter);
 
-        if (allCandidates.size() < 2) {
-            throw new IllegalArgumentException("Too few matching coordinates ("
-                    + allCandidates.size() + "). Wrong region imported?");
-        }
-        if (timeSteps.size() < 2) {
-            throw new IllegalStateException("Coordinates produced too few time steps "
-                    + timeSteps.size() + ", gpxList:" + gpxList.size());
+        // now look up the entries up in the graph:
+        final QueryGraph queryGraph = new QueryGraph(routingGraph).setUseEdgeExplorerCache(true);
+        List<QueryResult> allQueryResults = new ArrayList<>();
+        for (List<QueryResult> qrs: queriesPerEntry)
+            allQueryResults.addAll(qrs);
+        queryGraph.lookup(allQueryResults);
+
+        logger.info("================= Query results =================");
```

Comment:
Actually logger.debug should be called instead of logger.info for the logging in MapMatching.java. Then this logging can be easily configured via log config and we don't need a switch anymore (unless we are worried about performance). 

IMO, the added logging statements make sense because they help to figure out what is really going on during map matching. This is always useful when a trace is not map matched as expected. I don't think that more tests would replace this need for logging.